nebula3/gclient/net/SessionPool.py

```diff
@@ -0,0 +1,463 @@
+            # remove the session from the pool if it is invalid
+            self._active_sessions.remove(session)
+            raise e
+
+    def close(self):
+        """log out all sessions and close all connections
+
+        :return: void
+        """
+        with self._lock:
+            for session in self._idle_sessions:
+                session._sign_out()
+                session._connection.close()
+            for session in self._active_sessions:
+                session._sign_out()
+                session._connection.close()
+            self._idle_sessions.clear()
+            self._close = True
+
+    def get_ok_servers_num(self):
+        """get the number of the ok servers
+
+        :return: int
+        """
+        count = 0
+        for addr in self._addresses_status.keys():
+            if self._addresses_status[addr] == self.S_OK:
+                count = count + 1
+        return count
+
+    def _get_services_status(self):
+        msg_list = []
+        for addr in self._addresses_status.keys():
+            status = 'OK'
+            if self._addresses_status[addr] != self.S_OK:
+                status = 'BAD'
+            msg_list.append('[services: {}, status: {}]'.format(addr, status))
+        return ', '.join(msg_list)
+
+    def update_servers_status(self):
+        """update the servers' status"""
+        for address in self._addresses:
+            if self.ping(address):
+                self._addresses_status[address] = self.S_OK
+            else:
+                self._addresses_status[address] = self.S_BAD
+
+    def ping_sessions(self):
+        """ping all sessions in the pool"""
+        with self._lock:
+            for session in self._idle_sessions:
+                session.execute(r'RETURN "SESSION PING"')
+
+    def _get_idle_session(self):
+        """get a valid session from the pool idle list.
+
+        :return: Session
+        """
+        with self._lock:
+            if len(self._idle_sessions) > 0:
+                return self._idle_sessions.pop(0)
+            elif len(self._active_sessions) < self._configs.max_size:
+                return self._new_session()
+            else:
+                raise NoValidSessionException(
+                    'The total number of sessions reaches the pool max size {}'.format(
+                        self._configs.max_size
+                    )
+                )
+
+    def _new_session(self):
+        """construct a new session with the username and password in the pool.
+            also, the session is bound to the space specified in the configs.
+
+        :return: Session
+        """
+        self._pos = (self._pos + 1) % len(self._addresses)
+        addr = self._addresses[self._pos]
+        if self._addresses_status[addr] == self.S_OK:
+            if self._ssl_configs is None:
+                connection = Connection()
+                try:
+                    connection.open(addr[0], addr[1], self._configs.timeout)
```

Comment:
**In case one of the GraphD in the list is unavailable, it's raised here directly, if 1 of 10 graphD is down, this behavior may not be the best.**

What if we do things more accurately on connection-level cases?
- create a local _pos from self._pos
- identify if it's actually a connection issue to graphD:
    - init a connection `retry_timer` (default it's `len(self._addresses)`)
    - update `self._address_status[addr]` to `self.S_BAD`; `retry_timer--`
    - retry with `_pos + 1 % len` until `retry_timer <= 0`
        - if no workable attempts raise E with the reasoning of no GraphD is available now
        - if there are workable ones, print log/warning of S_BAD addresses

Reply:
Fixed.

Reply:
Oh, the interval_check is in seconds, right? I didn't realize it's checked so frequently, thus mentioning to do an update of self.S_BAD. Thanks